lib/scripting/grammar/grammar.ts

```diff
@@ -0,0 +1,289 @@
+								tokens.push(' ');
+							}
+						}
+					}
+					tokens.push(token.text);
+					prevToken = token;
+					separator = false;
+				}
+			},
+		});
+
+		return output;
+	}
+
+	public transpile(script: string): Program {
+		const stmts: Statement[] = [];
+
+		const formattedScript = this.format(script);
+
+		const vbsAst = this.parser.getAST(formattedScript, this.GRAMMAR_TARGET_PROGRAM);
+
+		if (vbsAst === null) {
+			throw new Error('Unable to transpile script:\n\n' + formattedScript);
+		}
+
+		const postProcessors = this.postProcessors;
+
+		dashAst(vbsAst, {
+			leave(node: ESIToken, parent: ESIToken) {
+				if (node.type === 'Program') {
+					for (const child of node.children) {
+						if (child.estree) {
+							if (!Array.isArray(child.estree)) {
+								stmts.push(child.estree);
+							} else {
+								stmts.push(...child.estree);
+							}
+						}
+					}
+				} else {
+					let estree: any = null;
+					/**
+					 * Loop through all registered post processors until an estree
+					 * is returned. If no post processors can handle the node, and
+					 * the node has a child, copy it's estree. This will allow estrees
+					 * to get propagated to parents, and we wont need to have
+					 * post-processors for every single rule in the grammar.
+					 */
+					for (const postProcessor of postProcessors) {
+						estree = postProcessor(node);
+						if (estree) {
+							break;
+						}
+					}
+					if (estree !== null) {
+						node.estree = estree;
+					} else if (node.children[0]) {
+						node.estree = node.children[0].estree;
+					}
+				}
+			},
+		});
+
+		return program(stmts);
+	}
+
+	public vbsToJs(script: string): string {
+		return generate(this.transpile(script));
+	}
+
+	/**
+	 * To keep the grammar readable, keywords are defined matching the case
+	 * from MS documentation. Each keyword is stored in a lookup table, and
+	 * then turned into a case insensitive version. For example, 'ByVal' will
+	 * become (B|b)(Y|y)(V|v)(A|a)(L|l). During the format process, all
+	 * keywords will be standardized using the lookup table.
+	 */
+
+	private setKeywords(grammar: string) {
+		const startIndex = grammar.indexOf(this.TOKEN_TERMINAL_KEYWORDS) + this.TOKEN_TERMINAL_KEYWORDS.length;
+		const endIndex = grammar.indexOf('\n', startIndex);
+
+		for (let keyword of grammar.substr(startIndex, endIndex - startIndex).split('|')) {
+			keyword = keyword.trim().slice(1, -1);
+			this.keywords[keyword.toLowerCase()] = keyword;
+		}
+	}
+
+	private addCaseInsensitiveKeywords(grammar: string): string {
+		const caseInsensitiveKeywords: string[] = [];
+
+		const startIndex = grammar.indexOf(this.TOKEN_TERMINAL_KEYWORDS) + this.TOKEN_TERMINAL_KEYWORDS.length;
+		const endIndex = grammar.indexOf('\n', startIndex);
+
+		for (const key of Object.keys(this.keywords)) {
+			let caseInsensitiveKeyword = '';
+			for (const letter of key) {
+				caseInsensitiveKeyword += "('" + letter.toUpperCase() + "'|'" + letter.toLowerCase() + "')";
```

Comment:
using Template literals here would improve readability

Reply:
Cool. I will need to look into template literals

Reply:
It's pretty easy:
```ts
caseInsensitiveKeyword += `('${letter.toUpperCase()}'|'${letter.toLowerCase()}')`;
```